### PR TITLE
TestProxy: Better handling of generic signatures in exception message

### PR DESCRIPTION
**Pull request by h2000**
See comments for proposed (==>) code which works with with TestProxy 

```java
test("A proxy handling generics works with proposed code templates",() -> {
			final Dependency<String> proxy = TestProxy.proxy(Dependency.class, new Object() {
				// ==> public <T> T generic(T x0){}
				public <T> T generic(T x0){ return x0; }

				// ==> public U baz(List<U> x0){}
				// works: public <X> X baz(List<X> x0){ return x0.get(0);}
				// works: public String baz(List<String> x0){ return x0.get(0);}
				// does not work public  int baz(List<Integer> x0){ return x0.get(0);}
				public <X> X baz(List<X> x0){ return x0.get(0);}

			});
			Expect.expectString(proxy.generic("abc")).isEqual("abc");
			Expect.expectString(proxy.baz(Arrays.asList("abc"))).contains("abc");
		});
```

### CODE REVIEW

**Comment by lefou** [comment]
Thanks for your follow up to #31. Could you please address my previous review comments:

* [ ] parameter naming
* [ ] unbound type variables
* [ ] unused proxy code

The reason, why I still prefer to use the lowest base type in the proposed signature for generics is, that we actually take no measure to ensure that casts to tighter type-bounds are safe.

Also, please have a look at the following two existing test cases:

```java
		test("A proxy handing an erased generic arg method should succeed", () -> {
			final Generic<String> dep = TestProxy.proxy(Generic.class, new Object() {
				@SuppressWarnings("unused")
				public boolean genericArg(Object arg1) {
					return true;
				}
			});
			expectTrue(dep.genericArg("1"));
		});

		test("A proxy handing an non-erased generic arg method should fail", () -> {
			final Generic<String> dep = TestProxy.proxy(Generic.class, new Object() {
				@SuppressWarnings("unused")
				public boolean genericArg(String arg1) {
					return true;
				}
			});
			intercept(UnsupportedOperationException.class,
					"(?s).*\\Qpublic boolean genericArg(T x0)\\E.*",
					() -> dep.genericArg("1"));
		});
```

The first uses `Object` for `T` and succeeds. The second uses `String` for `T` and must fail. So why does your change replaces the proposed signature from `public boolean genericArg(Object object1)` to `public boolean genericArg(T x0)`? Where does `T` come from?

**Comment by h2000** [comment]
Addressed the Codacy Quality points. (unused proxy code)

File: src/test/java/de/tobiasroeser/lambdatest/proxy/ExampleProxyTest.java [hunk 1]
```diff
@@ -107,22 +119,105 @@ public boolean genericArg(String arg1) {
...
+		section("A proxy with with missing implementation should print a nice (copy 'n paste -able) method signature",
+				() -> {
+					test("case: int foobar(String s1, int i2)",
+							() -> {
+								final Dependency dep = TestProxy.proxy(Dependency.class, new Object() {
+									@SuppressWarnings("unused")
+									public String hello() {
+										return "Hello Proxy!";
+									}
+								});
```
Comment (lefou) [inline src/test/java/de/tobiasroeser/lambdatest/proxy/ExampleProxyTest.java]: The delegate object isn't needed at all in this and also the next test cases.

```java
final Dependency dep = TestProxy.proxy(Dependency.class);
```

would be enough.

File: src/test/java/de/tobiasroeser/lambdatest/proxy/ExampleProxyTest.java [hunk 2]
```diff
@@ -107,22 +119,105 @@ public boolean genericArg(String arg1) {
...
+					test("case: U baz(List<U> arg1)",
+							() -> {
+								final Dependency dep = TestProxy.proxy(Dependency.class, new Object() {
+									@SuppressWarnings("unused")
+									public String hello() {
+										return "Hello Proxy!";
+									}
+								});
+								intercept(UnsupportedOperationException.class,
+										"(?s).*public U baz\\(List<U> list0\\).*",
```
Comment (lefou) [inline src/test/java/de/tobiasroeser/lambdatest/proxy/ExampleProxyTest.java]: When the proposed method is copied as-is, the compiler will most probably fail for the unknown `U`.

```java
new TestProxy(Dependency.class, new Object() {
  public U baz(List<U> list0) {
    return list0.get(0);
  }
});
```

File: src/test/java/de/tobiasroeser/lambdatest/proxy/ExampleProxyTest.java (hunk 1 above)
Comment (h2000) [reply src/test/java/de/tobiasroeser/lambdatest/proxy/ExampleProxyTest.java]: done.

File: src/test/java/de/tobiasroeser/lambdatest/proxy/ExampleProxyTest.java (hunk 2 above)
Comment (h2000) [reply src/test/java/de/tobiasroeser/lambdatest/proxy/ExampleProxyTest.java]: the comiler would be happy when we add <U> in front of it...

Comment (lefou) [reply src/test/java/de/tobiasroeser/lambdatest/proxy/ExampleProxyTest.java]: In most situations you can't and won't control the interface. You want a cheap implementation, thus the TestProxy. I guess, you can also add the `<U>` in the class, but haven't tested it. Nevertheless, you need to detect situations, in which the type parameter is defined non-locally, e.g. in the interface. Or do we do this already? Also, as the compiler makes no guarantees, you can simply replace it with `Object`, but that is a matter of taste, I guess.

**Comment by h2000** [comment]
Added if interfaces has type param: 
```
   ==>  public U baz(List<U> list0) { ... }  
OR ==>  public Object baz(List list0) { ... } 
```

**Comment by lefou** [comment]
Merged. Thank you!

I had to replace some internals to avoid Java8 API, as we also support Java7 runtimes.

**Comment by h2000** [comment]
Thx alot.

**Pull request merged**